### PR TITLE
Vetoshnikova Ekaterina Lab4 Var1

### PR DESCRIPTION
**Pull request by Veto1234-ii**

### CODE REVIEW

File: mlir/lib/Transforms/lab4/vetoshnikova_ekaterina/convertToCeildiv.cpp [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+private:
+  void replaceCeildiv(Operation *op, bool isSigned) {
+
+    OpBuilder builder(op);
+    Location loc = op->getLoc();
+    Value a = op->getOperand(0);
+    Value b = op->getOperand(1);
+
+    Value one = builder.create<arith::ConstantIntOp>(
+        loc, 1, builder.getIntegerType(32));
```
Comment (aobolensk) [inline mlir/lib/Transforms/lab4/vetoshnikova_ekaterina/convertToCeildiv.cpp]: Please add tests for other integer sizes

**Review by aobolensk — CHANGES_REQUESTED** [review]

File: mlir/lib/Transforms/lab4/vetoshnikova_ekaterina/convertToCeildiv.cpp [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+           "into calculation following the rule: ceildiv(a, b) = (a + b - 1) / "
+           "b ";
+  }
+
+  void runOnOperation() override {
+    getOperation()->walk([&](Operation *op) {
+      if (auto ceilDivUI = dyn_cast<arith::CeilDivUIOp>(op)) {
+        replaceCeildiv(ceilDivUI, false);
+      } else if (auto ceilDivSI = dyn_cast<arith::CeilDivSIOp>(op)) {
+        replaceCeildiv(ceilDivSI, true);
```
Comment (m-ly4) [inline mlir/lib/Transforms/lab4/vetoshnikova_ekaterina/convertToCeildiv.cpp]:
```
      if (isa<arith::CeilDivUIOp, arith::CeilDivSIOp>(op)) {
        replaceCeildiv(op);
```

File: mlir/lib/Transforms/lab4/vetoshnikova_ekaterina/convertToCeildiv.cpp [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+      }
+    });
+  }
+
+private:
+  void replaceCeildiv(Operation *op, bool isSigned) {
```
Comment (m-ly4) [inline mlir/lib/Transforms/lab4/vetoshnikova_ekaterina/convertToCeildiv.cpp]: Please implement this function without additional argument `isSigned`.

**Review by m-ly4 — CHANGES_REQUESTED** [review]

**Review by aobolensk — APPROVED** [review]

File: mlir/lib/Transforms/lab4/vetoshnikova_ekaterina/convertToCeildiv.cpp [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    Location loc = op->getLoc();
+    Value a = op->getOperand(0);
+    Value b = op->getOperand(1);
+
+    Value one = builder.create<arith::ConstantIntOp>(loc, 1, a.getType());
+    Value add = builder.create<arith::AddIOp>(loc, a, b);
+    Value sub = builder.create<arith::SubIOp>(loc, add, one);
+    Value div;
+
+    if (sub.getType().isSignlessInteger() && b.getType().isSignlessInteger()) {
```
Comment (m-ly4) [inline mlir/lib/Transforms/lab4/vetoshnikova_ekaterina/convertToCeildiv.cpp]: Ok, this check is appropriate.
Bit you could do the same:
```
if (isa<arith::CeilDivSIOp>(op)) {
```

**Review by m-ly4 — APPROVED** [review]

**Pull request merged**